v2/internal/crdmanagement/crds.go

```diff
@@ -0,0 +1,340 @@
+	existing []apiextensions.CustomResourceDefinition,
+	goal []apiextensions.CustomResourceDefinition,
+	comparators ...func(a apiextensions.CustomResourceDefinition, b apiextensions.CustomResourceDefinition) bool,
+) map[string]apiextensions.CustomResourceDefinition {
+
+	// Just invert the comparators and call FindMatchingCRDs
+	invertedComparators := make([]func(a apiextensions.CustomResourceDefinition, b apiextensions.CustomResourceDefinition) bool, 0, len(comparators))
+	for _, c := range comparators {
+		c := c
+		invertedComparators = append(
+			invertedComparators,
+			func(a apiextensions.CustomResourceDefinition, b apiextensions.CustomResourceDefinition) bool {
+				return !c(a, b)
+			})
+	}
+
+	return m.FindMatchingCRDs(existing, goal, invertedComparators...)
+}
+
+func (m *Manager) ApplyCRDs(ctx context.Context, path string, podNamespace string, patterns []string) error {
+	goalCRDs, err := m.LoadOperatorCRDs(path, podNamespace)
+	if err != nil {
+		return err
+	}
+	m.logger.V(Info).Info("Goal CRDs", "count", len(goalCRDs))
+
+	existingCRDs, err := m.ListOperatorCRDs(ctx)
+	if err != nil {
+		return err
+	}
+	m.logger.V(Info).Info("Existing CRDs", "count", len(existingCRDs))
+
+	goalCRDsWithDifferentVersion := m.FindNonMatchingCRDs(existingCRDs, goalCRDs, VersionEqual)
+	goalCRDsWithDifferentSpec := m.FindNonMatchingCRDs(existingCRDs, goalCRDs, SpecEqual)
+
+	// The same CRD may be in both sets, but we don't want to apply twice, so combine the sets
+	crdsToApply := make(map[string]apiextensions.CustomResourceDefinition)
+	for name, crd := range goalCRDsWithDifferentVersion {
+		crdsToApply[name] = crd
+	}
+	for name, crd := range goalCRDsWithDifferentSpec {
+		crdsToApply[name] = crd
+	}
+
+	if len(crdsToApply) > 0 {
+		for _, crd := range crdsToApply {
+			crd := crd
+
+			_, isDifferentVersion := goalCRDsWithDifferentVersion[crd.Name]
+			_, isDifferentSpec := goalCRDsWithDifferentSpec[crd.Name]
+			m.logger.V(Verbose).Info("Found CRD in need of update", "CRD", crd.Name, "SpecDifferent", isDifferentSpec, "VersionDifferent", isDifferentVersion)
+
+			toApply := &apiextensions.CustomResourceDefinition{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: crd.Name,
+				},
+			}
+			result, err := controllerutil.CreateOrUpdate(ctx, m.kubeClient, toApply, func() error {
+				resourceVersion := toApply.ResourceVersion
+				*toApply = crd
+				toApply.ResourceVersion = resourceVersion
+
+				return nil
+			})
+			if err != nil {
+				return errors.Wrapf(err, "failed to apply CRD %s", crd.Name)
+			}
+
+			m.logger.V(Debug).Info("Successfully applied CRD", "name", crd.Name, "result", result)
+		}
+
+		// If we make it to here, we have successfully updated all the CRDs we needed to. We need to kill the pod and let it restart so
+		// that the new shape CRDs can be reconciled.
+		m.logger.V(Status).Info("Restarting operator pod after updating CRDs", "count", len(crdsToApply))
+		os.Exit(0) // TODO: Should this be nonzero?
```

Comment:
I think zero is appropriate - this isn't a failure condition, it's normal operation.

Reply:
Removed the TODO